#30 Feat : 로그인 - Interceptor, 로그아웃 추가

yunah7 opened the pull request:
## 요약
- Interceptor을 통한 로그인 유지/자동 로그인
- 로그아웃 기능 추가
- 세션 ID를 properties 상수로 추가
- SignUpReq 어노테이션 추가

## 변경 내용

**1. AutoLoginInterceptor - 세션에 등록된 유저 검사**
```
@Component
@RequiredArgsConstructor
public class AutoLoginInterceptor implements HandlerInterceptor {

	private final UserService userService;

	@Override
	public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) throws
		Exception {
		HttpSession httpSession = request.getSession(false);
		if (httpSession != null) {
			Integer user = (Integer)httpSession.getAttribute(UserConstant.SESSION_ID);
			if (user != null) {
				return true;
			}
		}

		throw new NoMatchedUserException(NO_LOGIN_INFORMATION);
	}
```
- HandlerInterceptorAdapter 클래스가 deprecated된 관계로 인터페이스인 Handlerceptor 사용
  - 인터페이스 구현 방법에 따라 사용하지는 않지만, postHandle, afterCompletion 메소드도 override
- 세션의 유무와, 세션에 저장된 유저(USER 테이블의 ID 반환)의 유무를 검사하는 로직 작성


**2. WebConfig - Interceptor 등록**
```
@Configuration
@EnableWebMvc
@RequiredArgsConstructor
public class WebConfig implements WebMvcConfigurer {

	private final AutoLoginInterceptor autoLoginInterceptor;

	@Override
	public void addInterceptors(InterceptorRegistry registry) {
		registry.addInterceptor(autoLoginInterceptor)
			.order(1)
			.addPathPatterns("/*")
			.excludePathPatterns("/users/sign-up", "/users/sign-in");
	}

}
```
- order(1) 메소드를 통해 우선순위 부여
- addPathPatterns 메소드를 통해 모든 페이지에 해당 interceptor 등록
- excludePathPatterns 메소드를 통해 로그인, 회원가입 페이지 interceptor 적용 제외


**3. UserController, LoginService - 로그아웃 추가**
- UserController
```
	@PostMapping("/logout")
	public void signOut() {
		loginService.logout();
	}
```
- LoginService
```
	public void logout() {
		unsetLoginSession();
	}
```
- signup/signin/signout으로 통일할지, signup/login/logout으로 통일할지 미정


**4. UserConstant - 상수 추가**
```
    public static int MAX_LOGIN_TRY_COUNT;
    public static int MAX_PASSWORD_VALIDITY_PERIOD;
    public static String SESSION_ID;
```
- SESSION_ID를 service에서 String으로 넘기는 부분을 상수로 넘기도록 수정


**5. SignUpRes - 에러 해결**
```
@Getter
@Builder
@AllArgsConstructor
@NoArgsConstructor
```
- Builder 어노테이션만을 사용할 경우, 생성자 관련 에러가 생겨, 생성자를 만드는 어노테이션 추가


## Issue number/URL
#30

uri010 committed:
Pull Request 템플릿 추가

uri010 committed:
버그 이슈 템플릿 추가

uri010 committed:
기능추가 이슈 템플릿 추가

uri010 committed:
기능추가 템플릿 수정

내용 입력이 안되어 있어서 수정함

yunah7 committed:
Merge branch 'develop' of https://github.com/Shared-Calendar/Shared-Calendar

yunah7 committed:
#28 Bug - 실행 에러

yunah7 committed:
#30 feat : 로그인 - Interceptor 및 로그아웃

HandlerInterceptor을 상속받은 AutoLoginInterceptor 클래스를 통해 로그인 검사
WebConfig 클래스로 AutoLoginInterceptor 클래스를 회원가입, 로그인 페이지를 제외한 모든 페이지에 적용
로그인 상태가 아닐 때, 로그인이 필요한 페이지에 접속 시 예외 처리
로그아웃 기능 추가
세션 아이디를 properties 내 변수로 추가

sunwootest commented on src/main/java/study/sharedcalendar/constant/ErrorCode.java:
```diff
@@ -1,25 +1,27 @@
-    /* 404 NOT_FOUND */
-    NO_MATCHING_USER_ID(HttpStatus.NOT_FOUND, "아이디와 일치하는 유저가 없습니다."),
-    NO_MATCHING_USER_PASSWORD(HttpStatus.NOT_FOUND, "비밀번호가 일치하지 않습니다."),
+	/* 403 FORBIDDEN */
+	INACTIVE_USER(HttpStatus.FORBIDDEN, "휴면 계정입니다."),
+	EXCEEDED_LOGIN_ATTEMPTS(HttpStatus.FORBIDDEN, "로그인 횟수가 초과했습니다. 비밀번호를 재설정해주세요."),
+	EXCEEDED_PASSWORD_VALIDITY_PERIOD(HttpStatus.FORBIDDEN, "비밀번호 변경 주기가 지났습니다. 비밀번호를 재설정해주세요."),
+	/* 404 NOT_FOUND */
+	NO_MATCHING_USER_ID(HttpStatus.NOT_FOUND, "아이디와 일치하는 유저가 없습니다."),
+	NO_MATCHING_USER_PASSWORD(HttpStatus.NOT_FOUND, "비밀번호가 일치하지 않습니다."),
+	NO_LOGIN_INFORMATION(HttpStatus.NOT_FOUND, "로그인 상태가 아닙니다."),
```
약간의 팁을 드리자면 HttpStatus를 static import하면 코드가 더 간결해집니다 :)

sunwootest replied:
더 나아가서 프로그램의 규모가 커지면 모든 도메인들의 에러메시지를 한 클래스에 담기엔 관리도 어렵고 알아보기도 힘들겁니다. 그럴 땐 어떻게 해야되는지 고민하는 것도 좋을 것 같네요 :)

yunah7 replied:
static import로 설정하면 에러가 떠, 일단 작성해둔 코드들을 마저 올리고 찾아보겠습니다!

sunwootest replied:
인텔리제이 단축키를 활용해보세요. static import를 적용할 수 있는 코드에 커서를 옮기고 옵션 엔터치면 됩니다. 윈도우 명령어는 찾아봐야겠네요 :)

uri010 replied:
전에 저도 HttpStatus를 static import로 설정하려고 했는데INTERNAL_SERVER_ERROR(HttpStatus.INTERNAL_SERVER_ERROR,...)해당 코드에서 설정한 코드 이름과 HttpStatus의 필드 값이 같아서 앞에 HttpStatus를 적어줬었습니다.
ErrorCode를 수정하거나 해당 코드에만 HttpStatus를 지정하는 식으로 시도해보면 될거 같아요..!

sunwootest commented on src/main/java/study/sharedcalendar/interceptor/AutoLoginInterceptor.java:
```diff
@@ -0,0 +1,49 @@
+import org.springframework.web.servlet.HandlerInterceptor;
+import org.springframework.web.servlet.ModelAndView;
+
+import lombok.RequiredArgsConstructor;
+import study.sharedcalendar.constant.UserConstant;
+import study.sharedcalendar.exception.NoMatchedUserException;
+import study.sharedcalendar.service.UserService;
+
+@Component
+@RequiredArgsConstructor
+public class AutoLoginInterceptor implements HandlerInterceptor {
+
+	private final UserService userService;
+
+	@Override
+	public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) throws
+		Exception {
+		HttpSession httpSession = request.getSession(false);
+		if (httpSession != null) {
+			Integer user = (Integer)httpSession.getAttribute(UserConstant.SESSION_ID);
+			if (user != null) {
+				return true;
+			}
+		}
+
+		throw new NoMatchedUserException(NO_LOGIN_INFORMATION);
+	}
+
+	@Override
+	public void postHandle(HttpServletRequest request, HttpServletResponse response, Object handler,
+		ModelAndView modelAndView) throws Exception {
+		HandlerInterceptor.super.postHandle(request, response, handler, modelAndView);
+	}
+
+	@Override
+	public void afterCompletion(HttpServletRequest request, HttpServletResponse response, Object handler,
+		Exception ex) throws Exception {
+		HandlerInterceptor.super.afterCompletion(request, response, handler, ex);
+	}
+}
```
tab의 크기가 좀 커진 것 같은데요? 이제 협업단계이니 컨벤션을 맞추는 것이 좋아보입니다. 기업의 컨벤션 소스들이 인터넷에 공개되어 있으니 참고해보세요 :)

yunah7 replied:
네이버 코드 컨벤션으로 맞춰두었는데, IDE에서와는 달리 깃허브에서 조금 더 크게 보이는 것 같습니다..! 다시 한번 찾아보겠습니다 :)

sunwootest commented on src/main/java/study/sharedcalendar/interceptor/AutoLoginInterceptor.java:
```diff
@@ -0,0 +1,49 @@
+
+	private final UserService userService;
+
+	@Override
+	public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) throws
+		Exception {
+		HttpSession httpSession = request.getSession(false);
+		if (httpSession != null) {
+			Integer user = (Integer)httpSession.getAttribute(UserConstant.SESSION_ID);
+			if (user != null) {
+				return true;
+			}
+		}
+
+		throw new NoMatchedUserException(NO_LOGIN_INFORMATION);
+	}
```
전에도 한번 말씀드렸지만 우리가 접속하는 url에 만약 로그인 기능이 필요하지 않은 메소드가 존재한다면 매번 예외를 던질 수 있겠죠? 이 문제점을 인식하고 개발을 진행하시면 좋을 것 같습니다 :)

yunah7 replied:
해당 부분을 염두하여, URL을 가능한 세분화해 작성하고 있습니다 ! 참고하겠습니다 😸

sunwootest commented on src/main/java/study/sharedcalendar/service/LoginService.java:
```diff
@@ -1,63 +1,68 @@
+	private final HttpSession httpSession;
+
+	public void login(LoginReq loginReq) {
+		User user = userMapper.findLoginUser(loginReq);
```
로그인한 유저의 정보는 세션에서 가져오는 것이 가장 정확하겠죠? 세션이 어떤식으로 사용되고 저장되며 구분되는지 이해하면 좋겠습니다 :)

yunah7 replied:
해당 부분을 제가 작성하지는 않아 정확한 의도를 말씀드리긴 어렵지만,  로그인 시 받아온 입력값과 매칭되는 DB값을 찾는 부분인 것 같습니다 !

sunwootest replied:
지금 확인해보니 그러네요 ㅎㅎ맞게 작성된 것 같습니다

sunwootest commented on src/main/java/study/sharedcalendar/service/LoginService.java:
```diff
@@ -1,63 +1,68 @@
+	private final EncryptionService encryptionService;
+	private final HttpSession httpSession;
+
+	public void login(LoginReq loginReq) {
+		User user = userMapper.findLoginUser(loginReq);
+
+		if (user == null) {
+			throw new NoMatchedUserException(NO_MATCHING_USER_ID);
+		}
+
+		if (!encryptionService.isMatch(loginReq.getPassword(), user.getPassword())) {
+			loginTryCountCheck(user.getTryCount());
+			userMapper.incrementLoginTryCount(user.getId());
+			throw new AuthorizationException(NO_MATCHING_USER_PASSWORD);
+		}
+
+		if (!user.isActivate()) {
+			throw new AuthorizationException(INACTIVE_USER);
+		}
+
+		if (userMapper.getPasswordDateDiff(user.getId()) >= userConstant.MAX_PASSWORD_VALIDITY_PERIOD) {
+			throw new AuthorizationException(EXCEEDED_PASSWORD_VALIDITY_PERIOD);
+		}
+
+		loginTryCountCheck(user.getTryCount());
+		userMapper.initLoginTryCount(user.getId());
+		setLoginSession(user.getId());
+	}
```
하나의 mapper는 하나의 service에만 대응하는 것이 추후에 유지보수 하기 좋습니다. 빈 의존관계가 섞이면 나중에 곤란해지거든요 :)

yunah7 replied:
LoginMapper 클래스로 분리하겠습니다 :)

sunwootest commented on src/main/java/study/sharedcalendar/service/LoginService.java:
```diff
@@ -1,63 +1,68 @@
+		setLoginSession(user.getId());
+	}
+
+	public void logout() {
+		unsetLoginSession();
+	}
+
```
이 정도는 메소드로 분리 안해도 될 것 같습니다 :)

uri010 commented on src/main/java/study/sharedcalendar/dto/SignUpReq.java:
```diff
@@ -1,30 +1,31 @@
+
 @Getter
 @Builder
+@AllArgsConstructor
+@NoArgsConstructor
```
생성자가 꼭 있어야 하나요?

uri010 replied:
이전에 제가 짠 코드에서는 Constructor 관련 어노테이션 없이 Builder 어노테이션만 써도 사용 가능했었습니다. 확인 부탁드려요!

sunwootest replied:
이 부분은 jackson이 어떻게 json을 자바코드로 직렬화 하는지 알아보면 알 수 있습니다. 아마 이렇게 생성자 두개 만들어야 될겁니다 :)

uri010 replied:
찾아보니 다른 생성자가 있는데 Builder를 사용할 경우 AllArgsConstructor를 사용해야 한다고 합니다. Builder만 사용하는 경우 생성자 관련 어노테이션은 필요 없을 거 같습니다.

yunah7 replied:
Builder 어노테이션만을 사용할 경우, 실행 자체는 가능하지만, json 데이터를 Object 객체로 deserialize 할 때, 생성자의 부재로 인한 InvalidDefinitionException 에러가 발생해 두 어노테이션을 적용해두었습니다 ! Json관련 다른 어노테이션으로도 해결할 수 있다고 보긴했는데, 에러의 원인이 직관적이라, Constructor관련 어노테이션을 사용했습니다.

uri010 replied:
아하 설명 감사합니다😄

yunah7 committed:
#30 fix : 로그인 - logout 메소드, LoginMapper 수정

 관심사 분리를 위한 LoginMapper 클래스 분리
 로그아웃 메소드 수정

yunah7 commented:
## 변경 사항
- LoginMapper 클래스 분리
```
@Mapper
public interface LoginMapper {
	User findLoginUser(LoginReq loginReq);

	void incrementLoginTryCount(int id);

	void initLoginTryCount(int id);

	int getPasswordDateDiff(int id);
}
```
login과 관련된 메소드들은 클래스를 분리해 작성하였습니다.

- Logout 메소드 변경
```
	public void logout() {
		httpSession.invalidate();
	}
```

yunah7 committed:
#30 fix : xml 파일 분화

sunwootest commented:
LoginMapper에 해당하는 내용들은 모두 userMapper로 옮길 수 있을 것 같네요. 하나의 테이블에 굳이 2개의 mapper를 연결시킬 필요는 없습니다. LoginService가 MemberService를 참조해도 될 것 같아 보이네요 :)

yunah7 committed:
#30 fix : 로그인 - Mapper 참조 변경

UserMapper 내 메소드를 UserService에서 사용해 UserSevice 객체를 통해 LoginService에서 UserMapper을 사용하도록 변경

yunah7 commented:
## 변경 사항

- UserService를 통해 LoginService가 UserMapper을 사용하도록 변경
```

	public User findLoginUser(LoginReq loginReq) {
		return userMapper.findLoginUser(loginReq);
	}

	public void incrementLoginTryCount(int id) {
		userMapper.incrementLoginTryCount(id);
	}

	public void initLoginTryCount(int id) {
		userMapper.initLoginTryCount(id);
	}

	public int getPasswordDateDiff(int id) {
		return userMapper.getPasswordDateDiff(id);
	}
```

sunwootest commented:
일단은 이 정도 수준으로만 객체지향 설계를 하도록 하죠. 제가 생각했던 방향은 LoginService가 UserService에 의존해서 userMapper의 정보를 가져오는거였습니다 :)

sunwootest reviewed (approved):
approve 드립니다. 다른 팀원도 확인하고 approve 주세요 :)

uri010 reviewed (approved):
approve 드리겠습니다!
